datastore/src/main/java/org/jboss/pnc/datastore/repositories/ProductMilestoneRepositoryImpl.java

```diff
@@ -55,4 +75,77 @@ public long countBuiltArtifactsInMilestone(Integer id) {
+                .from(DeliverableAnalyzerOperation.class);
+        Join<DeliverableAnalyzerOperation, ProductMilestone> productMilestones = deliverableAnalyzerOperations
+                .join(DeliverableAnalyzerOperation_.productMilestone);
+        Join<DeliverableAnalyzerOperation, DeliverableAnalyzerReport> deliverableAnalyzerReports = deliverableAnalyzerOperations
+                .join(DeliverableAnalyzerOperation_.report);
+        Join<DeliverableArtifact, Artifact> artifacts = deliverableAnalyzerReports
+                .join(DeliverableAnalyzerReport_.artifacts)
+                .join(DeliverableArtifact_.artifact);
+        Join<Artifact, TargetRepository> targetRepositories = artifacts.join(Artifact_.targetRepository);
+
+        List<Selection<?>> selects = Lists.newArrayList(
+                artifacts.get(Artifact_.id),
+                artifacts.get(Artifact_.deployPath),
+                targetRepositories.get(TargetRepository_.repositoryType));
+        selects.addAll(getArtifactPresenceInMilestonesCaseColumns(cb, productMilestones, milestoneIds));
+
+        query.multiselect(selects);
+        query.where(
+                getMavenOrNpmArtifactsInMilestonesPredicate(
+                        cb,
+                        targetRepositories,
+                        productMilestones,
+                        deliverableAnalyzerReports,
+                        milestoneIds));
+
+        TypedQuery<Tuple> typedQuery = entityManager.createQuery(query);
+        return typedQuery.getResultList();
+    }
+
+    private List<Selection<Boolean>> getArtifactPresenceInMilestonesCaseColumns(
+            CriteriaBuilder cb,
+            Path<ProductMilestone> productMilestones,
+            List<Integer> milestoneIds) {
+        // maps milestone IDs into columns with boolean value whether Artifact was delivered in that Milestone
+        return milestoneIds.stream().map(milestoneId -> {
+            CriteriaBuilder.SimpleCase<Integer, Boolean> simpleCase = cb
+                    .selectCase(productMilestones.get(ProductMilestone_.id));
+            simpleCase.when(milestoneId, true);
+            simpleCase.otherwise(false);
+
+            return simpleCase;
+        }).collect(Collectors.toList());
+    }
+
+    private Predicate getMavenOrNpmArtifactsInMilestonesPredicate(
+            CriteriaBuilder cb,
+            Path<TargetRepository> targetRepositories,
+            Path<ProductMilestone> productMilestones,
+            Path<DeliverableAnalyzerReport> deliverableAnalyzerReports,
+            List<Integer> milestoneIds) {
+        Predicate isMavenOrNpmArtifact = targetRepositories.get(TargetRepository_.repositoryType)
+                .in(RepositoryType.MAVEN, RepositoryType.NPM);
+
+        Expression reportLabels = deliverableAnalyzerReports.get(DeliverableAnalyzerReport_.labels).as(String.class);
+        Predicate notFromScratchAnalysis = cb
```

Comment:
i needed the same predicate here:
https://github.com/project-ncl/pnc/pull/4141/files#diff-37fdf2be118f8c59981d031c3a0a8ad5f57f774576e93f191533c29b9aa8f1ebR556

any idea where to extract it? there are some predicate classes but follow different structure than this predicate

Reply:
I'd add it to predicate class or create the predicate if it's missing. Doesn't matter that this is more generic but the class name categorizes where the predicate belongs.